extensions.html

```diff
@@ -1145,6 +617,38 @@ <h3>
         </ol>
       </section>
     </section>
+    <section>
+      <h2>
+        Partial <dfn>GamepadHapticActuator</dfn> Interface
+      </h2>
+      <p>
+        This partial interface supplements the {{GamepadHapticActuator}}
+        interface described in the main
+        <a href="https://w3c.github.io/gamepad/">Gamepad spec</a>.
+      </p>
+      <pre class="idl">
+        [Exposed=Window]
+        partial interface GamepadHapticActuator {
+          Promise&lt;boolean&gt; pulse(double value, double duration);
+        };
+      </pre>
+      <dt>
+        <dfn>pulse()</dfn> method
+      </dt>
+      <dd>
+        <p>
+          <code>pulse()</code> applies a <var>value</var> to the actuator for
+          <var>duration</var> milliseconds. The <var>value</var> passed to
+          <code>pulse()</code> is clamped to limits defined by the actuator
+          type. The returned Promise will resolve <code>true</code> once the
+          pulse has completed.
+        </p>
+        <p>
+          Repeated calls to <code>pulse()</code> override the previous
```

Comment:
What does override mean here? What happens to the p1?

```
p1  = ...pulse();
p2  = ...pulse()
```

Does p1 get rejected?

Reply:
I think the only implementation of `pulse` is in Meta Quest Browser. We should find out how it behaves and write the spec to match that implementation as closely as possible.

I looked at the original PR (https://github.com/w3c/gamepad/pull/32) and didn't see any discussion of whether the promise returned by pulse should ever do anything but resolve to true. Maybe @toji remembers?

Personally, I would want the promise returned by `pulse()` to resolve to true only if the effect played for the entire duration without being pre-empted. I expect apps will want to chain multiple calls to `pulse()` to make more complex effects. If one of the pulses is pre-empted, apps should be able to detect that and avoid playing the subsequent chained pulses.

So, I would say `p1` should be resolved with false after the second call to `pulse()`. We shouldn't reject in this case because pre-empting an effect is not an error.

Reply:
@cabanier ☝️...

Reply:
Ok, but if pulse() is Meta Quest only, that also puts it the "feature at risk". We should note that in the spec too, unless we expect to get a second implementation. 

Is there a plan for a second implementation?

Reply:
Actually, this is basically an incubation, so we don't need to block on this. 

But we should figure out if anyone will actually support pulse() or not and if not, we should consider dropping it.

Reply:
Broadly agree with @marcoscaceres that we should evaluate whether or not `pulse()` is actually desired going forward. I'll also throw up the @cabanier bat signal here to get Meta's take on it.

Re: the intent behind the `pulse()` promise resolution...
![84_Years](https://github.com/w3c/gamepad/assets/805273/b349d91c-52f3-44a1-b065-e3f193857784)

It _has_ been a while, so I don't recall the details of the conversation around this API. But I do recall having conversations with the TAG in regards to similarly shaped APIs at around the same time. Specifically the [WebXR `isSessionSupported()` API](https://immersive-web.github.io/webxr/#xrsystem-interface), which also resolves to a boolean value. 

Originally we were going to have the promise resolve with void if the session type was supported and reject otherwise but the guidance we received was that rejection should only happen in the case of something truly exceptional, where the intent of the method called could not be fulfilled. This is especially true when you consider that many promises are used with async/await patterns, and the only way to observe the rejection is a try/catch. That's fairly disruptive. In the case of `isSessionSupported()` the question being asked (can I start this kind of session) could always be answered yes or no, so resolving with a boolean made sense and the method never rejects.

With that framing in mind, the fact that this resolves to a boolean rather than void tells me that the boolean was intended to communicate information. It feels to my like the intent was "resolve when the requested pulse is finished, with true if it was allowed to complete and false if it was pre-empted." I wouldn't expect it to reject unless perhaps the pulse simply couldn't be rendered by the device haptics at all? Even then it seems possibly more ergonomic to never reject.

Whether or not that's what the Meta implementation does is an entirely different story.

Reply:
yeah, I agree that not rejecting makes sense. 

I'm still hopeful pulse can go away though... unless it's standardized across different controllers, it's still quite ambiguous about what it actually means or how it would be represented/felt (duration, strength, amplitude, etc.).

Reply:
What do you think that we should do now? Should we remove it or keep it in the extensions document?

Reply:
yeah, let's keep it in the extensions document.

Reply:
> Ok, but if pulse() is Meta Quest only, that also puts it the "feature at risk". We should note that in the spec too, unless we expect to get a second implementation.
> Is there a plan for a second implementation?

I'm fairly sure that this was also implemented in the Pico browser. @bkardell , was this added to Wolvic?

Reply:
> With that framing in mind, the fact that this resolves to a boolean rather than void tells me that the boolean was intended to communicate information. It feels to my like the intent was "resolve when the requested pulse is finished, with true if it was allowed to complete and false if it was pre-empted." I wouldn't expect it to reject unless perhaps the pulse simply couldn't be rendered by the device haptics at all? Even then it seems possibly more ergonomic to never reject.
> 
> Whether or not that's what the Meta implementation does is an entirely different story.

I checked the Quest implementation and it's running `pulse` through the same codepath as `playEffect` after clamping.
Looking through the code, I can't see where this would return false except for cases where the controller doesn't have haptics (ie hands)

Reply:
That doesn't sound great because it seems to deviate away from playEffect()'s behavior. If they are using the same code path, then we should really do playEffect("pulse") as a followup.

Reply:
> > Ok, but if pulse() is Meta Quest only, that also puts it the "feature at risk". We should note that in the spec too, unless we expect to get a second implementation.
> > Is there a plan for a second implementation?
> 
> I'm fairly sure that this was also implemented in the Pico browser. @bkardell , was this added to Wolvic?

The version of Gecko that is shipped with the latests releases of wolvic (116) does implement Pulse()

Reply:
Right, if we want to include it then we should do it as a followup with more detail about making it somewhat more interoperable (i.e., what does pulse actually do across various gamepads and how do we give them a consistent feel...). At the same time, we should add the behavior to be more controllable via vibrationActuator, because .pulse() is super crude.